Sources/SourceKitLSP/Swift/MacroExpansionReferenceDocumentURLData.swift

```diff
@@ -0,0 +1,133 @@
+
+  package init(
+    macroExpansionEditRange: Range<Position>,
+    sourceFileURL: URL,
+    selectionRange: Range<Position>,
+    bufferName: String
+  ) {
+    self.sourceFileURL = sourceFileURL
+    self.selectionRange = selectionRange
+    self.bufferName = bufferName
+    self.macroExpansionEditRange = macroExpansionEditRange
+  }
+
+  private var displayName: String {
+    if let macroExpansionEditRange {
+      "L\(macroExpansionEditRange.lowerBound.line + 1)C\(macroExpansionEditRange.lowerBound.utf16index + 1)-L\(macroExpansionEditRange.upperBound.line + 1)C\(macroExpansionEditRange.upperBound.utf16index + 1).swift"
+    } else {
+      ""
+    }
+  }
+
+  private var queryItems: [URLQueryItem]? {
+    get {
+      let queryItems = [
+        URLQueryItem(name: Parameters.sourceFilePath, value: sourceFileURL.path(percentEncoded: false)),
+        URLQueryItem(name: Parameters.fromLine, value: String(selectionRange.lowerBound.line)),
+        URLQueryItem(name: Parameters.fromColumn, value: String(selectionRange.lowerBound.utf16index)),
+        URLQueryItem(name: Parameters.toLine, value: String(selectionRange.upperBound.line)),
+        URLQueryItem(name: Parameters.toColumn, value: String(selectionRange.upperBound.utf16index)),
+        URLQueryItem(name: Parameters.bufferName, value: bufferName),
+      ]
+
+      return queryItems
+    }
+  }
+
+  package var url: URL {
+    get throws {
+      var components = URLComponents()
+      components.scheme = "sourcekit-lsp"
+      components.host = "swift-macro-expansion"
+      components.path = "/\(displayName)"
+      components.queryItems = queryItems
+
+      guard let url = components.url else {
+        throw ReferenceDocumentURLError(
+          description: "Unable to create URL for macro expansion reference document"
+        )
+      }
+
+      return url
+    }
+  }
+
+  package init(from url: URL) throws {
+    guard url.scheme == "sourcekit-lsp" else {
+      throw ReferenceDocumentURLError(description: "Invalid scheme for reference document")
+    }
+
+    guard url.host(percentEncoded: false) == "swift-macro-expansion" else {
+      throw ReferenceDocumentURLError(
+        description: "Invalid document type for macro expansion reference document"
+      )
+    }
+
+    guard let urlComponents = URLComponents(string: url.absoluteString),
+      let queryItems = urlComponents.queryItems,
+      let sourceFilePath = queryItems.first { $0.name == Parameters.sourceFilePath }?.value,
+      let fromLine = Int(queryItems.first { $0.name == Parameters.fromLine }?.value ?? ""),
+      let fromColumn = Int(queryItems.first { $0.name == Parameters.fromColumn }?.value ?? ""),
+      let toLine = Int(queryItems.first { $0.name == Parameters.toLine }?.value ?? ""),
+      let toColumn = Int(queryItems.first { $0.name == Parameters.toColumn }?.value ?? ""),
+      let bufferName = queryItems.first { $0.name == Parameters.bufferName }?.value
```

Comment:
Nitpicky, but I’d do `last` because usually the semantics are that a repeated statement of the same parameter overrides previous parameters.